### PR TITLE
feat/basic auth hybrid mode (take 2)

### LINKED ISSUE #5649: Basic-auth plugin does not work in hybrid mode
### Summary
Basic-auth plugin does not work in hybrid mode - when I try to create consumer credentials, changes on control plane are not propagated to data plane. According to [this post](https://discuss.konghq.com/t/kong-2-hybrid-mode-and-plugins-compatibility/5565) it should work for auth plugins, and for key-auth it actually works fine. Details has been [described here](https://discuss.konghq.com/t/basic-auth-plugin-in-hybrid-mode/5689).

### Steps To Reproduce

1. Install Kong in hybrid mode (1 control plane, 1 data plane)
2. Create a consumer on control plane
3. Create a basic-auth credentials for that consumer
4. Try to use the credentials on data plane

### Additional Details & Logs
Config_hash checksum (`curl -X GET localhost:8001/clustering/status`) changes when creating a consumer or key-auth credentials, but it does not change when creating basic-auth credentials.

- Kong version (`2.0.1`)
- Kong error logs - no errors
- Operating system - Docker

### PR DESCRIPTION
**Pull request by kikito**
Alternative name for this PR: feat/declarative config _transform.

This change fixes #5649 (basic-auth credentials not working on hybrid mode).

It is an alternative to #5792 where the proposed fix involves adding a new pseudo-field called `encrypted_password` to the entities.

The main drawback of that approach is that it requires manually changing all the entities which have transformations. Moreover, the change seems to be nearly identical in all cases: making a pseudo-field which in effect deactivates the transformation.

On this take, we add a new parameter to the declarative config file itself. The field is called `_transform`. When this field is present and set to `false`, the transformations are not executed when exporting. If the field is not present, or if it is, but it's set to `true`, the transformations are executed (which is the current behavior).

This approach does not require individual changes to any entities - a single flag in the top-level yaml is enough.

However, since this flag is new, the yaml version needed to be bumped up.

Also, this approach assumes that *on export*, we never want to do transformations (`_transform` is `false` on all exports). If we ever encounter a situation in which we *do* want to have transformations even after a straight db export, we will need to find a different way.

### CODE REVIEW

File: kong/db/declarative/init.lua [hunk 1]
```diff
@@ -241,7 +275,8 @@ function declarative.export_from_db(fd)
   end
 
   fd:write(declarative.to_yaml_string({
-    _format_version = "1.1",
+    _format_version = "2.0",
```
Comment (hutchic) [inline kong/db/declarative/init.lua:278]: does this require an update to docs?

File: kong/templates/kong_yml.lua [hunk 2]
```diff
@@ -12,6 +12,15 @@ return [[
 
 _format_version: "1.1"
```
Comment (bungle) [inline kong/templates/kong_yml.lua:13]: Should this be updated?

**Comment by hbagdi** [comment]
How hard is it to expose this feature to the Admin API as well?
decK currently sync passwords back to Kong after a reset because there is no way to skip the 'hashing' part in the Admin API.
Is adding a query string parameter to skip hashing when the password is already hashed a good idea and something we should add?

cc @hishamhm

File: kong/db/declarative/init.lua (hunk 1 above)
Comment (kikito) [reply kong/db/declarative/init.lua]: Indeed. I have switched the version to 1.2 everywhere.

File: kong/templates/kong_yml.lua (hunk 2 above)
Comment (kikito) [reply kong/templates/kong_yml.lua]: Yes, I have updated it to 1.2.

**Comment by kikito** [comment]
@hbagdi Some parts of this PR will make it easier to implement something similar (a special kong header?) on the Admin API. In particular, the DAO functions now accept a `options.transform` setting which can be used. I think it would't be very difficult to add such a flag (that falls out of the scope of this particular PR).

As a side note: my previous take on this problem (#5792) did in fact work on both the Admin API and declarative config. However a single "global switch" was preferred.

**Comment by hbagdi** [comment]
Got it. Added https://github.com/Kong/kong/issues/5925 to track it.
Thanks @kikito !

File: kong/db/declarative/init.lua [hunk 3]
```diff
@@ -351,7 +389,17 @@ function declarative.get_current_hash()
 end
 
 
-function declarative.load_into_cache(entities, hash, shadow_page)
+-- dc_table format:
+--   {
+--     _format_version: 1.1,
```
Comment (bungle) [inline kong/db/declarative/init.lua:394]: And this

File: kong/db/declarative/init.lua [hunk 4]
```diff
@@ -144,12 +157,26 @@ function Config:parse_string(contents, filename, accept, old_hash)
 end
 
 
+-- first return item: a table with the following format:
+--   {
+--     _format_version: 1.1,
```
Comment (bungle) [inline kong/db/declarative/init.lua:162]: This too.

File: kong/db/declarative/init.lua [hunk 5]
```diff
@@ -78,14 +78,27 @@ function Config:parse_file(filename, accept, old_hash)
...
```
Comment (bungle) [inline kong/db/declarative/init.lua:83]: Should we update these as well?

File: spec/01-unit/01-db/01-schema/11-declarative_config/02-process_auto_fields_spec.lua [hunk 6]
```diff
@@ -23,6 +23,7 @@ describe("declarative config: process_auto_fields", function()
         config = DeclarativeConfig:process_auto_fields(config, "select", false)
         assert.same({
           _format_version = "1.1",
```
Comment (bungle) [inline spec/01-unit/01-db/01-schema/11-declarative_config/02-process_auto_fields_spec.lua:25]: What about these?

Comment (kikito) [reply spec/01-unit/01-db/01-schema/11-declarative_config/02-process_auto_fields_spec.lua:25]: Those are fine. The 1.2 can understand and accept 1.1-format. It is ok to test that if you put 1.1 into the process_auto_fields function, you get a 1.1 version back.

**Comment by hishamhm** [comment]
> How hard is it to expose this feature to the Admin API as well?

@hbagdi According to this Pr, the boolean whether transformations are triggered or not is an attribute the declarative config input itself, so by definition it is already supported by the Admin API. You just need to pass `_transform` as false when sending in the data to `/config` the same way you already pass `_format_version`.

Or am I missing something? @kikito ?

**Comment by hbagdi** [comment]
> @hbagdi According to this PR, the boolean whether transformations are triggered or not is an attribute the declarative config input itself, so by definition it is already supported by the Admin API. You just need to pass _transform as false when sending in the data to /config the same way you already pass _format_version.

I should have been clearer. By Admin API, I meant via the `/consumers/:username/basic-auth`  endpoint.
The problem that we need to solve is that users use decK to export the configuration but can't import basic-auth credentials back. That results in Kong re-hashing them. If there is a way to avoid re-hashing, that solves that problem.
decK currently emits a warning when someone tries to "sync" the passwords but it still trips users because this is the only field that has this behavior in all of Kong and Kong Enterprise.

**Pull request merged**